hash_practice/exercises.py

```diff
@@ -22,8 +49,29 @@ def valid_sudoku(table):
         Each element can either be a ".", or a digit 1-9
         The same digit cannot appear twice or more in the same 
         row, column or 3x3 subgrid
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: O(n**2)
+        Space Complexity: O(1)
     """
-    pass
-
+    # Note from Lizet to instructor: I know this doesn't pass the last 2 tests, but I'm not sure why.
+    # test cases but submitting as is since this last function is optional
+    # iterate through each row
+    for row in table:
+        # iterate through each element in the row
+        for element in row:
+            # if the element is not a "."
+            if element != ".":
+                # if the element is not a digit 1-9
+                if element not in "123456789":
+                    return False
+                # if the element is a digit 1-9
+                else:
+                    # if the element is in the same row
+                    if row.count(element) > 1:
+                        return False
+                    # if the element is in the same column
+                    if [row[i] for i in range(9)].count(element) > 1:

```

Comment:
I know this is an optional problem, but here's some hints. Note that `row` is going to remain the same single row during each iteration of the loop, so `[row[i] for i in range(9)]` is just going to return a copy of `row` (given our assumption that each row has 9 elements).

To get each column you'd have to do something like `[table[i][curr_col] for i in range(9)]`, where `curr_col` is the current column index, but you would likely have to change your code to keep track of what that is, and go through the indices rather than the rows something like this:

```
for curr_row in range(len(table)):
    for curr_col in range(table[curr_row]):
        # do your checks here
```